### ISSUE TITLE
Make it easier to build with binary caches

### ISSUE THREAD

**Issue by mightybyte** [haskell-nix/hnix#314]
Using binary caches is crucial to working with nix.  If you don't, you can easily end up with builds that take many hours.  Because of this, getting set up to work on a new project that uses nix often requires you to edit the `nix.conf` file and add the appropriate cache and public key.  This may be ok in some cases but is sometimes too cumbersome.  HNix should make it easy to add new binary caches on a more granular basis.  Perhaps via a command line option, or possibly in a per-project config file.

**Comment by shlevy** [comment]
Binary caches are fundamentally a trust issue, so they can't be automatic. In any case, they're a property of the nix-daemon, not the evaluator.

**Comment by ryantrinkle** [comment]
Agreed regarding trust, but I think being able to programmatically *check*
whether a cache is working would be good. For most reflex-platform users,
an error is better than a 12 hour build

**Comment by shlevy** [comment]
@ryantrinkle Agreed. FWIW it seems like such functionality can be independent of `hnix` and simply depend directly on `hnix-store` once it's ready, though for usability purposes it's probably nice to have them in the same program namespace.

**Comment by shlevy** [comment]
I wonder if we should have a separate `porcelain` or `exes` project/package...

**Comment by jwiegley** [comment]
@shlevy Yes to separate porcelain, but when the project has matured a bit.  Nix still makes  working on two projects simultaneously unnecessarily painful.

**Comment by mightybyte** [comment]
As a fundamental trust issue, maybe it's not desirable to have projects turn on caches transparently, but it sure would be nice to be able to say `nix-build --cache http://mycache.mydomain.com` and have that just work.  If the cache is explicitly requested by the user like that, it seems to me that you wouldn't need to explicitly specify a public key since you're tacitly saying that you trust the things coming from that domain.  But if there is any objection to that, you can always add a `--cache-key abc123...` argument.

**Comment by domenkozar** [comment]
I'm working on something to make it really easy to add binary caches (have nix.conf parser, etc).

**Comment by shlevy** [comment]
@mightybyte you can already specify substituters on the command line... nix-build --substituters https://example.com/nix-cache?trusted=1

**Comment by mightybyte** [comment]
@shlevy Oh nice!  I hadn't seen that.  @jwiegley perhaps you can try that approach and see if it works as an alternative to what I hacked up for `hnix-web-repl`?

**Comment by mightybyte** [comment]
@shlevy FYI, the man pages for `nix-build` and `nix-shell` do not mention the `--substituters` option.

**Comment by shlevy** [comment]
@mightybyte I don't know if it's documented anywhere, but in 2.0 nix.conf settings all get a corresponding command line flag. The existence of `--substituters` comes from the existence of the `substituters` nix.conf option.

**Comment by jwiegley** [comment]
@mightybyte Giving it a try now. The old way of doing it wasn't working on Travis, causing the ghcjs build to time out every attempt.

**Comment by domenkozar** [comment]
@jwiegley happy to help setup https://cachix.org/ - it should get CI times down to ~3min.

**Comment by jwiegley** [comment]
@domenkozar Let's do it. It should require simply modifying `build.sh`.
